### PR TITLE
Document types support in CLI

### PR DESCRIPTION
**Pull request by zdutta**
*Description of changes:*

Introducing document type support. Changes made to help docs and error raised if short hand syntax is used with a document type input. Additional testing dependency and service modifications to test changes in theory: https://github.com/zdutta/botocore/tree/document-types.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: awscli/clidocs.py [hunk 1]
```diff
@@ -451,7 +453,13 @@ def _do_json_example(self, doc, argument_model, stack):
             doc.style.dedent()
             doc.write('}')
         elif argument_model.type_name == 'structure':
-            self._doc_input_structure_members(doc, argument_model, stack)
+            if argument_model.is_document_type:
+                doc.write("This value is an JSON document that can have arbitrary content.")
+                doc.style.new_line()
+                doc.write("For more information on JSON document values visit: <insert section name in user guide>")
```
Comment (mtdowling) [inline awscli/clidocs.py:459]: are document type values provided to the CLI using a JSON file or something?

File: awscli/argprocess.py [hunk 2]
```diff
@@ -379,6 +379,16 @@ def _handle_special_cases(self, cli_argument, value, service_id,
             # {"Value": "shutdown"}
             return {'Value': value}
 
+        # Additional handling of special cases where a structure is
+        # a document type. Shorthand syntax is not supported and
+        # should therefore err out before attempt to parse.
+        if model.type_name == 'structure' and model.is_document_type:
```
Comment (mtdowling) [inline awscli/argprocess.py:385]: Why no shorthand support for document types? What if the value the user needs to provide are scalars like `"string"` or `true` or `10`?

File: awscli/clidocs.py [hunk 3]
```diff
@@ -451,7 +453,13 @@ def _do_json_example(self, doc, argument_model, stack):
             doc.style.dedent()
             doc.write('}')
         elif argument_model.type_name == 'structure':
-            self._doc_input_structure_members(doc, argument_model, stack)
+            if argument_model.is_document_type:
+                doc.write("This value is an JSON document that can have arbitrary content.")
```
Comment (mtdowling) [inline awscli/clidocs.py:457]: s/an/a

If you're taking JSON here in the CLI, then you might say "JSON value" instead to clarify that any JSON value is accepted.

**Comment by kyleknap** [comment]
I just pushed an update on this. The commit message details most of the changes I made. With regards to reviewing this, I'd recommend:

1. Going over the functional tests that I added to understand the end to end behavior with regards to document types.
2. If you want to try it yourself, I have a sample service model: https://gist.github.com/kyleknap/097638826d2041b283442dbad23257ae to manually test out. You can start trying the interface by downloading the model and running:

  ```
  aws configure add-model --service-model files://path/to/downloaded/service-2.json
  aws doctype describe-resource help  # Pulls up sample man page

 aws configure set cli_history enabled
 aws doctype describe-resource --doc-type-param '{"foo": "bar"}'  # Sample doc type usage
 aws history show --include API_CALL  # Shows serialization of parameter
  ```

File: awscli/clidocs.py [hunk 4]
```diff
@@ -393,6 +400,18 @@ def _add_webapi_crosslink(self, help_command):
         doc.style.external_link(title="AWS API Documentation", link=link)
         doc.writeln('')
 
+    def _add_note_for_document_types_if_used(self, help_command):
+        if operation_uses_document_types(help_command.obj):
+            help_command.doc.style.new_paragraph()
+            help_command.doc.writeln(
+                '``%s`` uses document type values. Document types follow the '
+                'JSON data model where valid values are: strings, numbers, '
+                'booleans, null, arrays, and objects. For command input,'
```
Comment (zdutta) [inline awscli/clidocs.py:409]: Missing a space after the last comma

**Comment by kyleknap** [comment]
@stealthycoin @zdutta I updated based on feedback. It should be good to look at again.

**Review by stealthycoin — APPROVED** [review]
Tested it out, worked as expected.

**Comment by zdutta** [comment]
Looks good to me.

**Pull request merged**